packages/mds-ingest-migration-processor/processor.ts

```diff
@@ -30,49 +32,147 @@ import {
   convert_v1_0_0_vehicle_event_to_v1_1_0,
   Device_v0_4_1,
   Device_v1_1_0,
+  Nullable,
   Telemetry,
+  UUID,
   VehicleEvent_v0_4_1,
   VehicleEvent_v1_1_0
 } from '@mds-core/mds-types'
+import { asArray } from '@mds-core/mds-utils'
 import { cleanEnv, str } from 'envalid'
 
 const { SOURCE_TENANT_ID, TENANT_ID } = cleanEnv(process.env, { SOURCE_TENANT_ID: str(), TENANT_ID: str() })
 
 type MigrationEntityType = 'device' | 'event' | 'telemetry'
-const MigrationTopic = (tenant: string, entityType: MigrationEntityType) => `${tenant}.${entityType}`
+
+const MigrationSourceTopic = (entityType: MigrationEntityType) => `${SOURCE_TENANT_ID}.${entityType}`
 
 const MigrationDataSource: <MigrationSourceEntity>(entity: MigrationEntityType) => StreamSource<MigrationSourceEntity> =
   entity => {
-    const topic = MigrationTopic(SOURCE_TENANT_ID, entity)
+    const topic = MigrationSourceTopic(entity)
     return KafkaSource(topic, { groupId: `${topic}.ingest-migration-processor` })
   }
 
+const MigrationErrorTopic = (entityType: MigrationEntityType) => `${TENANT_ID}.${entityType}.error`
+
 const MigrationErrorSink: <MigrationSourceEntity>(
   entity: MigrationEntityType
 ) => DeadLetterSink<MigrationSourceEntity> = entity => {
-  const topic = `${MigrationTopic(TENANT_ID, entity)}.error`
+  const topic = MigrationErrorTopic(entity)
   return KafkaSink(topic, {
     clientId: `${topic}.ingest-migration-processor`
   })
 }
 
-const DevicesMigrationProcessor = StreamProcessor<Device_v0_4_1 & { id: number }, Device_v1_1_0>(
+const MigratedFrom = (entityType: MigrationEntityType, migrated_from_id: number) => ({
+  migrated_from_source: MigrationSourceTopic(entityType),
+  migrated_from_version: '0.4.1',
+  migrated_from_id
+})
+
+interface MigratedDevice {
+  device: Device_v1_1_0
+  migrated_from: MigratedEntityModel
+}
+
+const IngestServiceMigratedDeviceSink = (): StreamSink<MigratedDevice> => () => {
+  const topic = MigrationSourceTopic('device')
+  return {
+    initialize: async () => undefined,
+    write: async message => {
+      try {
+        const [{ device, migrated_from }] = asArray(message)
+        const migrated = await IngestServiceClient.writeMigratedDevice(device, migrated_from)
+        if (migrated) {
+          logger.info(`Migrated device ${migrated.device_id}`)
+        }
+      } catch (error) {
+        logger.error(`Error migrating device`, { topic, message })
+      }
+    },
+    shutdown: async () => undefined
+  }
+}
+
+const DevicesMigrationProcessor = StreamProcessor<Device_v0_4_1 & { id: number }, MigratedDevice>(
```

Comment:
Could the `Recorded<>` util be used here?

Reply:
I don't think that represents the over the wire shape coming into the processor, but I did change to using `& IdentityColumn` instead of `& { id: number }`